Use walleterror for error handling in try_get_internal_address

reez opened the pull request:
I've just been looking thru what functions we currently use (or I use in a sample iOS project) that have Alpha3Error so I can turn those into more specific errors.

<!-- Erase any parts of this template not applicable to your Pull Request. -->

### Description

Adds `PersistenceError` and replaces the use of `Alpha3Error` with `PersistenceError ` in the [try_get_internal_address](https://docs.rs/bdk/1.0.0-alpha.6/bdk/wallet/struct.Wallet.html#method.try_get_internal_address) function. 

I added a unit test for `PersistenceError` as well.

### Notes to the reviewers

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

reez commented on bdk-ffi/src/wallet.rs:
```diff
@@ -1,6 +1,6 @@
 use crate::bitcoin::{OutPoint, PartiallySignedTransaction, Transaction};
 use crate::descriptor::Descriptor;
-use crate::error::{Alpha3Error, CalculateFeeError, WalletCreationError};
+use crate::error::{Alpha3Error, CalculateFeeError, WalletError};
```
It still uses Alpha3Error in other methods. Resolving conversation

thunderbiscuit commented:
Those pesky errors. So now this remaps a write error from a PersistBackend::Write error to a NewOrLoadError::Write.

I'm worried about the confusion that we're about to introduce as Rust errors are attempted to be mapped to the ffi layer, but can't be fully mapped either (see my PR on persistence) because some of the methods change or combine multiple Rust calls and must therefore combine the potentially returned errors.

Sorry I'm just venting here. But errors are a big mess right now.

thunderbiscuit commented:
I'm playing with the [`thiserror`](https://docs.rs/thiserror/latest/thiserror/) library to see if we can make these errors a little easier to work with. thiserror is a very common library in the Rust ecosystem, and provides I think a good way to simplify our code and make it easier to maintain. I added 2 commits to this PR to demonstrate a little bit how it might work.

@reez @notmandatory let me know what you think. If there is interest, I might actually go and attempt a refactor of our current set of errors using it as a separate PR.

Here is the gist of what it comes down to:

### Before
```rust
#[derive(Debug)]
pub enum WalletError {
    // Errors coming from the FileError enum
    Io {
        e: String,
    },
    InvalidMagicBytes {
        got: Vec<u8>,
        expected: Vec<u8>,
    },

    // Errors coming from the NewOrLoadError enum
    Descriptor,
    Write,
    Load,
    NotInitialized,
    LoadedGenesisDoesNotMatch,
    LoadedNetworkDoesNotMatch {
        expected: Network,
        got: Option<Network>,
    },
}

impl fmt::Display for WalletError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io { e } => write!(f, "io error trying to read file: {}", e),
            Self::InvalidMagicBytes { got, expected } => write!(
                f,
                "file has invalid magic bytes: expected={:?} got={:?}",
                expected, got,
            ),
            Self::Descriptor => write!(f, "error with descriptor"),
            Self::Write => write!(f, "failed to write to persistence"),
            Self::Load => write!(f, "failed to load from persistence"),
            Self::NotInitialized => {
                write!(f, "wallet is not initialized, persistence backend is empty")
            }
            Self::LoadedGenesisDoesNotMatch => {
                write!(f, "loaded genesis hash does not match the expected one")
            }
            Self::LoadedNetworkDoesNotMatch { expected, got } => {
                write!(f, "loaded network type is not {}, got {:?}", expected, got)
            }
        }
    }
}

impl std::error::Error for WalletError {}
```

### thiserror
```rust
#[derive(Debug, thiserror::Error)]
pub enum WalletError {
    #[error("IO error, this may mean that the file is too short")]
    Io,

    #[error("file has invalid magic bytes: expected={expected:?} got={got:?}")]
    InvalidMagicBytes { got: Vec<u8>, expected: Vec<u8> },

    #[error("error with descriptor")]
    Descriptor,

    #[error("failed to write to persistence")]
    Write,

    #[error("failed to load from persistence")]
    Load,

    #[error("wallet is not initialized, persistence backend is empty")]
    NotInitialized,

    #[error("loaded genesis hash does not match the expected one")]
    LoadedGenesisDoesNotMatch,

    #[error("loaded network type is not {expected}, got {got:?}")]
    LoadedNetworkDoesNotMatch { expected: Network, got: Option<Network> },
}
```

reez commented:
Good work @thunderbiscuit 

I think we should try `thiserror` out, the one thing I do want us to keep an eye on is the potential increase in compilation times due to the procedural macros of `thiserror`.

ACK f2d45457cec193a4368feb57cf3b048aecb23394

reez commented:
@thunderbiscuit I rebased this on `master` so I could get the newest `thiserror` changes and such. So the changes in this PR atm are:
- replaces the use of `Alpha3Error` with `WalletCreationError` in the [try_get_internal_address](https://docs.rs/bdk/1.0.0-alpha.8/bdk/wallet/struct.Wallet.html#method.try_get_internal_address) function.
- added a unit test to test all cases of the `WalletCreationError` enum

Wanted to get your feedback on what your thoughts on how this should shape up before it is mergeable? I didn't rename `WalletCreationError` to `WalletError` anymore, but I'm still using `WalletCreationError`'s `WriteError` even though that may still seem like a bit of an awkward thing, so just wanted to get your thoughts on where you think that Error return type for `try_get_internal_address` should be in your head.

thunderbiscuit commented:
After playing with it a bit, it felt to me like those were fairly distinct errors. I feel like this is initially hidden by the fact that they both use the persistence's `WriteError` (try_get_internal_address uses it outright, new_or_load more indirectly as a generic on the enum). But fundamentally I think the errors are closer to this:

- [try_get_internal_address](https://docs.rs/bdk/1.0.0-alpha.8/bdk/wallet/struct.Wallet.html#method.try_get_internal_address): `PersistenceBackend::WriteError`
- [new_or_load](https://docs.rs/bdk/1.0.0-alpha.8/bdk/wallet/struct.Wallet.html#method.new_or_load): `NewOrLoadError<PersistenceBackend::WriteError>`

Which makes me want to reduce it to the following two distinct errors:
- try_get_internal_address: `PersistenceBackend::WriteError`
- new_or_load: `NewOrLoadError::WriteError`

Open to discussing more of course.

reez commented:
> After playing with it a bit, it felt to me like those were fairly distinct errors. I feel like this is initially hidden by the fact that they both use the persistence's `WriteError` (try_get_internal_address uses it outright, new_or_load more indirectly as a generic on its own enum error, `NewOrLoadError`). Fundamentally I think the errors are closer to this:
> 
> * [try_get_internal_address](https://docs.rs/bdk/1.0.0-alpha.8/bdk/wallet/struct.Wallet.html#method.try_get_internal_address): `PersistenceBackend::WriteError`
> * [new_or_load](https://docs.rs/bdk/1.0.0-alpha.8/bdk/wallet/struct.Wallet.html#method.new_or_load): `NewOrLoadError<PersistenceBackend::WriteError>`
> 
> Which makes me want to reduce it to the following two distinct errors:
> 
> * try_get_internal_address: `PersistenceBackend::WriteError`
> * new_or_load: `NewOrLoadError::WriteError`
> 
> Open to discussing more of course.

Pushed some updates to this per your comments and our convo, I think its ready to go, but let me know if any other thoughts/updates, happy to change anything else if needed-

thunderbiscuit reviewed (approved):
ACK c6d17b77ec764003f5d44ce5f547de8ab9deeec8.
